src/main/scala/beam/agentsim/agents/vehicles/VehicleEnergy.scala

```diff
@@ -0,0 +1,209 @@
+            mutable.Map(gradePercentBin -> mutable.Map(numberOfLanesBin -> rate))
+      }
+    })
+    currentRateFilter.toMap.map {
+      case (speedInMilesPerHourBin, gradePercentMap) =>
+        speedInMilesPerHourBin -> gradePercentMap.toMap.map {
+          case (gradePercentBin, lanesMap) => gradePercentBin -> lanesMap.toMap
+        }
+    }
+  }
+
+  private def convertRecordStringToRange(recordString: String) = {
+    Range(recordString.replace(",", ":").replace(" ", ""))
+  }
+}
+
+class VehicleEnergy(
+  consumptionRateFilterStore: ConsumptionRateFilterStore,
+  linkToGradeRecordsIterableUsing: CsvParser => Iterable[Record]
+) {
+  private lazy val log = LoggerFactory.getLogger(this.getClass)
+
+  private val settings = new CsvParserSettings()
+  settings.detectFormatAutomatically()
+  private val csvParser = new CsvParser(settings)
+
+  type ConsumptionRateFilter = Map[Range, Map[Range, Map[Range, Double]]] //speed->(gradePercent->(numberOfLanes->rate))
+  private lazy val linkIdToGradePercentMap = loadLinkIdToGradeMapFromCSV
+  val conversionRateForJoulesPerMeterConversionFromGallonsPer100Miles = 746.86
+
+  def getFuelConsumptionEnergyInJoulesUsing(
+    fuelConsumptionDatas: IndexedSeq[BeamVehicle.FuelConsumptionData],
+    overallFallBack: IndexedSeq[BeamVehicle.FuelConsumptionData] => Double,
+    singleFallBack: BeamVehicle.FuelConsumptionData => Double
+  ): Double = {
+    val consumptionsInJoules: IndexedSeq[Double] = fuelConsumptionDatas
+      .flatMap(fuelConsumptionData => {
+        for {
+          rateInJoulesPerMeter <- getRateUsing(fuelConsumptionData, singleFallBack)
+          distance             <- fuelConsumptionData.linkLength
+        } yield rateInJoulesPerMeter * distance
+      })
+    if (consumptionsInJoules.nonEmpty) consumptionsInJoules.sum else overallFallBack(fuelConsumptionDatas)
+  }
+
+  private def getRateUsing(fuelConsumptionData: BeamVehicle.FuelConsumptionData,
+                           fallBack: BeamVehicle.FuelConsumptionData => Double): Option[Double] = {
+    val BeamVehicle.FuelConsumptionData(linkId, vehicleType, numberOfLanesOption, _, _, _, speedInMilesPerHourOption, _, _, _) =
+      fuelConsumptionData
+    val numberOfLanes: Int = numberOfLanesOption.getOrElse(0)
+    val speedInMilesPerHour: Double = speedInMilesPerHourOption.getOrElse(0)
+    val gradePercent: Double = linkIdToGradePercentMap.getOrElse(linkId, 0)
+    consumptionRateFilterStore.getPrimaryConsumptionRateFilterFor(vehicleType).flatMap(
+      consumptionRateFilterFuture => getRateUsing(consumptionRateFilterFuture,
+        numberOfLanes, speedInMilesPerHour, gradePercent)
+    ).orElse(
+      consumptionRateFilterStore.getSecondaryConsumptionRateFilterFor(vehicleType).flatMap(
+        consumptionRateFilterFuture => getRateUsing(consumptionRateFilterFuture,
+          numberOfLanes, speedInMilesPerHour, gradePercent)
+      )
+    ).orElse(Some(fallBack(fuelConsumptionData)))
```

Comment:
Here, this is causing the problems. The fall back is returning a value in units of energy (J) instead of a rate (J/m). You are essentially multiplying by the length of the road twice, leading to huge amounts of energy being consumed. Make the fall back return just the rate and we should be good.

Reply:
Thanks - can't believe I missed that - fresh eyes are always helpful!